github/ratelimit_test.go

```diff
@@ -0,0 +1,331 @@
+	"testing"
+	"time"
+
+	"github.com/src-d/metadata-retrieval/testutils"
+
+	"github.com/stretchr/testify/suite"
+)
+
+const defaultRateLimitReset = 1 * time.Second
+const defaultAbuseReset = 1 * time.Second
+const default403Reset = 2 * time.Second
+
+var testCases = map[string]response{
+	// regular answer
+	"/normal": response{
+		statusCode: http.StatusOK,
+		headers:    nil,
+		response:   apiResponse{Data: "success"},
+		err:        nil,
+	},
+	// proper ratelimit reset in defaultRateLimitReset
+	"/ratelimit_sleep": response{
+		statusCode: http.StatusOK,
+		headers:    rateLimitHeaders(defaultRateLimitReset),
+		response:   apiResponse{Data: "whatever"},
+		err:        nil,
+	},
+	// proper abuse reset in defaultRateLimitReset
+	"/abuse_sleep": response{
+		statusCode: http.StatusForbidden,
+		headers:    abuseHeaders(defaultAbuseReset),
+		response:   apiResponse{Data: "whatever"},
+		err:        nil,
+	},
+	// 403 with proper AbuseRateLimit body but without headers
+	"/abuse_no_headder": response{
+		statusCode: http.StatusForbidden,
+		headers:    nil,
+		response: apiResponse{
+			apiErrorResponse: apiErrorResponse{Message: "Found an abuse detection mechanism"},
+		},
+		err: nil,
+	},
+	// 403 without proper AbuseRateLimit body nor headers
+	"/forbidden_403": response{
+		statusCode: http.StatusForbidden,
+		headers:    nil,
+		response:   apiResponse{Data: "forbidden"},
+		err:        nil,
+	},
+	// returns a network error
+	"/network_error": response{
+		statusCode: http.StatusTeapot,
+		headers:    nil,
+		response:   apiResponse{},
+		err:        fmt.Errorf("network error"),
+	},
+}
+
+type response struct {
+	statusCode int
+	headers    func(time.Time) map[string]string
+	response   apiResponse
+	err        error
+}
+
+type apiResponse struct {
+	apiErrorResponse
+	Data string `json:"data,omitempty"`
+}
+
+func rateLimitHeaders(wait time.Duration) func(time.Time) map[string]string {
+	return func(when time.Time) map[string]string {
+		return map[string]string{
+			"X-RateLimit-Reset":     strconv.FormatInt(when.Add(wait).Unix(), 10),
+			"X-RateLimit-Remaining": "0",
+		}
+	}
+}
+
+func abuseHeaders(wait time.Duration) func(time.Time) map[string]string {
+	return func(when time.Time) map[string]string {
+		return map[string]string{
+			"Retry-After": strconv.FormatInt(int64(wait.Seconds()), 10),
+		}
+	}
+}
+
+func newResponse(content apiResponse, headers func(time.Time) map[string]string, statusCode int) (*http.Response, error) {
+	data, err := json.Marshal(content)
+	if err != nil {
+		return nil, err
+	}
+
+	respWriter := httptest.NewRecorder()
+	respWriter.WriteHeader(statusCode)
+	_, err = respWriter.Write(data)
+	if err != nil {
+		return nil, err
+	}
+
+	response := respWriter.Result()
+
+	if headers != nil {
+		for k, v := range headers(time.Now()) {
+			response.Header.Set(k, v)
+		}
+	}
+
+	return response, err
+}
+
+func newRequest(url string) *http.Request {
+	req, _ := http.NewRequest("GET", url, http.NoBody)
```

Comment:
it isn't important but let's use `POST` method here as all our requests are POST. Would be also nice to set some body and in the tests check that the server receives it correctly.

Reply:
good idea, done by 9fd372c